#146: Added boolean parameter use-node-name-as-external-address, defa…

henrikarro opened the pull request:
Pull request for issue #146; this adds the flag use-node-name-as-external-address to skip looking up the external IP using the /nodes resource.

fix #146

devOpsHazelcast commented:
Can one of the admins verify this patch?

henrikarro commented:
Just noticed I had trailing spaces on a few lines -- I turned off autosave cleanup since it uses different sort order for import statements -- so I pushed a new version. Only affects three lines in KubernetesClientTest and KubernetesConfigTest.

mesutcelik commented:
verify

googlielmo commented:
Hi @henrikarro thanks for the PR!
Could you please include user instructions in the README?

henrikarro commented:
I have updated `README.md` and also made the info message given when using `use-node-name-as-external-address` a bit less threatening. Also updated `.gitignore` with files that Checkstyle and FindBugs created, at least on my machine.

When updating `README.md`, I noticed that the `service-port` property is not documented. I didn't  do anything about this since it's unrelated to #146, but someone might want to look at it.

hasancelik commented:
> When updating README.md, I noticed that the service-port property is not documented. I didn't do anything about this since it's unrelated to #146, but someone might want to look at it.

Thanks for the catch @henrikarro 👍 I fixed it via https://github.com/hazelcast/hazelcast-kubernetes/commit/d08d198a18f2878839e8d1541984b3eb719638ba

googlielmo commented:
Thanks @henrikarro for addressing comments, approved!
I pushed a new commit to resolve conflicts.

leszko commented on src/main/java/com/hazelcast/kubernetes/KubernetesClient.java:
```diff
@@ -296,15 +299,20 @@ private static String extractZone(JsonObject nodeJson) {
-                        String nodeUrl = String.format("%s/api/v1/nodes/%s", kubernetesMaster, node);
-                        nodePublicIp = extractNodePublicIp(callGet(nodeUrl));
-                        cachedNodePublicIps.put(node, nodePublicIp);
+                        if (useNodeNameAsExternalAddress) {
```
Minor: consider extracting these 7 lines to a separate method.

henrikarro replied:
Annoying fact: when I extract that into a method, CheckStyle complains about `Total number of methods is 31 (max allowed is 30)`.

Any preferences about how to handle this?

leszko replied:
lol ;) Add `@SuppressWarnings` for now, we'll have to split this class into multiple classes, but let's not do it in this PR.

henrikarro replied:
Method extracted and `@SuppressWarnings("checkstyle:methodcount")` added.

leszko commented on src/test/java/com/hazelcast/kubernetes/KubernetesClientTest.java:
```diff
@@ -437,6 +441,26 @@ public void endpointsByNamespaceWithNodePublicIp() {
+    public void endpointsByNamespaceWithNodeName() {
+        // given
+        // reuse setup from another test case
+        endpointsByNamespaceWithNodePublicIp();
```
You call another test in this test, that is not right. Please extract the common parts to a method to reuse.

henrikarro replied:
True that. Extracted the JSON for the node port services into helper methods.

leszko commented on src/test/java/com/hazelcast/kubernetes/KubernetesClientTest.java:
```diff
@@ -56,12 +56,16 @@
                                            .willReturn(aResponse().withStatus(401).withBody("\"reason\":\"Forbidden\"")));
     }
 
+    private KubernetesClient newKubernetesClient(boolean useNodeNameAsExternalAddress) {
```
1. Make it static
2. Move this method down, since it's used in multiple places in this class.

henrikarro replied:
The method `newKubernetesClient` uses `wireMockRule`, which I believe cannot be made static, so unfortunately I cannot make the method static unless we provide `wireMockRule` as a parameter.

Which version do you prefer: static method with `wireMockRule` as parameter; or non-static method?

henrikarro replied:
It's only the port number we need from `wireMockRule`, so I can provide the port number as a parameter and make the method static. Then the caller needs to know how to find the port number...

henrikarro replied:
Left it non-static, moved it down to the other helper methods.

henrikarro commented:
A signed copy of the Contributor Agreement has been sent to cla@hazelcast.com.

leszko commented:
@henrikarro the PR looks fine. And thanks for signing CLA. One last thing, could you add your parameter `use-node-name-as-external-address` to `KubernetesConfig.toString()`?

henrikarro commented:
@leszko KubernetesConfig.toString has been updated.

leszko commented:
@henrikarro FYI: Your change is released in hazelcast-kubernetes 1.5.1.
